crates/ruff_server/src/edit.rs

```diff
@@ -29,6 +31,57 @@ pub enum PositionEncoding {
     UTF8,
 }
 
+/// A unique document ID, derived from a URL passed as part of an LSP request.
+/// This document ID can point to either be a standalone Python file, a full notebook, or a cell within a notebook.
+#[derive(Clone, Debug)]
+pub(crate) enum DocumentKey {
+    Notebook(PathBuf),
+    NotebookCell(lsp_types::Url),
+    Text(PathBuf),
+}
+
+impl DocumentKey {
+    /// Creates a document key from a URL provided in an LSP request.
+    pub(crate) fn from_url(url: &lsp_types::Url) -> Self {
+        if url.scheme() != "file" {
+            return Self::NotebookCell(url.clone());
+        }
```

Comment:
I'm assuming this works because VS Code chooses a different scheme (`notebook-cell` I think?) for Notebook cell documents. And, I thought of the same thing to do in `ruff-lsp` initially but it turns out that this is specific to VS Code and possibly wouldn't work across editors. The reason is that the spec mentions to not rely on the URI format:

> Cell text documents have an URI however servers should not rely on any format for this URI since it is up to the client on how it will create these URIs.
>
> Reference: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#notebookDocument_synchronization (Point 2)

This would work today because I think only VS Code supports Jupyter Notebook natively in the editor. But, it might not work for other editors (possibly Zed in the future).

The `ruff-lsp` implementation basically uses lookups to determine the document this URI belongs to. Basically, it'll check if there's a cell document with the URI and if it does, it's classified as a cell document. Similar lookup is done for Notebook document and at the end fall back to consider it as a text document.

Looking at the changes in this PR, I'm assuming something like that can be implemented by querying the `Index` although you'll have better context on how can this be implemented.

Reply:
Looking at the references of `DocumentKey::from_url`, it does seem that the references have access to the `Index`.

Reply:
It's also mentioned a bit below in the same reference section:

> Syncing the text content of a cell is relatively easy since clients should model them as text documents. However since the URI of a notebook cell’s text document should be opaque, servers can not know its scheme nor its path. However what is know is the notebook document itself. We therefore introduce a special filter for notebook cell documents:

Reply:
You're right, we could utilize the `Index` here instead of looking for things that may vary by implementation 👍